File: nim/src/nithellopkg/evaluate.nim
##[
  evaluate.nim
  ============

  module to contain various hueristic procs for the search algorithms.
]##

import strformat

from board import Board
from moves import getWeight
from player import Color
from util import `++`

type
  Scores* = object
    black*: int
    white*: int
    score*: int


proc calculateScoresDisc*(b: Board): Scores =
  var
    blackCount = 0
    whiteCount = 0

  for c in b.board:
    case c:
      of Black:
        ++blackCount
      of White:
        ++whiteCount
      of None:
        discard
  
  let score = blackCount - whiteCount

  result = Scores(black: blackCount, white: whiteCount, score: score)


proc calculateScoresWeight*(b: Board): Scores =
  var
    blackCount = 0
    whiteCount = 0

  for i, c in b.board:
    # var i = i
    case c:
      of Black:
        blackCount += getWeight(i)
      of White:
        whiteCount += getWeight(i)
      of None:
        discard

  let score = blackCount - whiteCount

  result = Scores(black: blackCount, white: whiteCount, score: score)


proc printResults*(s: Scores) =
  if s.black > s.white:
    echo "player black wins."
    echo fmt"black pieces: {s.black}"
    echo fmt"white pieces: {s.white}"
  elif s.white > s.black:
    echo "player white wins."
    echo fmt"black pieces: {s.black}"
    echo fmt"white pieces: {s.white}"
  else:
    echo "a tie occurred."
    echo fmt"black pieces: {s.black}"
    echo fmt"white pieces: {s.white}"
  

File: go/evaluate.go
package main

import (
	"fmt"
)

// Scores A struct to hold the scores of the board
type Scores struct {
	Black int
	White int
	Score int
}

func (b Board) calculateScoresDisc() Scores {
	// this winds up being used for max disc strats
	blackCount := 0
	whiteCount := 0

	for _, c := range b.board {
		switch {
		case c == black:
			blackCount++
		case c == white:
			whiteCount++
		}
	}

	r := blackCount - whiteCount

	return Scores{Black: blackCount, White: whiteCount, Score: r}
}

func (b Board) calculateScoresWeight() Scores {
	blackCount := 0
	whiteCount := 0

	for i, c := range b.board {
		switch {
		case c == black:
			blackCount += weights[i]
		case c == white:
			whiteCount += weights[i]
		}
	}

	r := blackCount - whiteCount

	return Scores{Black: blackCount, White: whiteCount, Score: r}
}

func printResults(s Scores) {
	switch {
	case s.Black > s.White:
		fmt.Println("Player Black wins.")
		fmt.Println("black pieces:", s.Black)
		fmt.Println("white pieces:", s.White)
		// fmt.Println("Player Black wins.")
	case s.Black < s.White:
		fmt.Println("Player White wins.")
		fmt.Println("black pieces:", s.Black)
		fmt.Println("white pieces:", s.White)
	case s.Black == s.White:
		fmt.Println("A tie occurred.")
		fmt.Println("black pieces:", s.Black)
		fmt.Println("white pieces:", s.White)
	}
}
